src/module.c

```diff
@@ -7896,6 +7917,62 @@ int RM_SubscribeToKeyspaceEvents(RedisModuleCtx *ctx, int types, RedisModuleNoti
     return REDISMODULE_OK;
 }
 
+void firePostKeySpaceJobs() {
+    /* Avoid propagation of commands. */
+    server.in_nested_call++;
+    while (listLength(modulePostKeyspaceJobs) > 0) {
+        listNode *ln = listFirst(modulePostKeyspaceJobs);
+        RedisModulePostKeyspaceJob *job = listNodeValue(ln);
+        listDelNode(modulePostKeyspaceJobs, ln);
+
+        RedisModuleCtx ctx;
+        moduleCreateContext(&ctx, job->module, REDISMODULE_CTX_TEMP_CLIENT);
+        selectDb(ctx.client, job->dbid);
+
+        job->callback(&ctx, job->pd);
+        if (job->free_pd) job->free_pd(job->pd);
+
+        moduleFreeContext(&ctx);
+        zfree(job);
+    }
+    server.in_nested_call--;
+}
+
+/* This API should only be executed within a key space notification callback and allow to
+ * react to the notification with some write operations.
+ *
+ * In general it is dangerous and highly discouraged to write inside key space notification callback.
+ * This is because the callback is executed synchronously in the middle of Redis logic
+ * (commands logic, eviction, expire). This means that writing inside a notification can cause
+ * data that held by Redis to be freed and cause Redis to crash when continue running his logic after.
+ *
+ * In order to still performing write actions as a reaction to key space notification, Redis gives
+ * `RM_AddPostNotificationJob` API. The api allows to register a job. When Redis will call this job
+ * the following condition are promised to be fulfilled:
+ * 1. It is safe to perform any write operation.
+ * 2. The job will be called atomically along side the notification (the notification effect and the
+ *    job effect will be replicated to the replication and the aof wrapped with multi exec).
```

Comment:
i think part of this warning is already covered in RM_SubscribeToKeyspaceEvents.
i don't think we should repeat ourselves, and also i wanna make this API more "general-purpose".
let's make sure all the KSN specific concerns are placed in RM_SubscribeToKeyspaceEvents, and make this one more generic

Reply:
Removed and add reference to `RM_SubscribeToKeyspaceEvents`. Regarding generalise the API, other then unlink (which I wrote [here](https://github.com/redis/redis/pull/11199#discussion_r979389759) I believe we should reconsider the chose we made), what other places do you think this API can be used?

Reply:
As discussed, API will be generalised but currently will be documented that it can be used on key space notifications.